src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage_Threading.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SQLite
+{
+    internal partial class SQLitePersistentStorage
+    {
+        /// <summary>
+        /// Use a <see cref="ConcurrentExclusiveSchedulerPair"/> to simulate a reader-writer lock.
+        /// Read operations are performed on the <see cref="ConcurrentExclusiveSchedulerPair.ConcurrentScheduler"/>
+        /// and writes are performed on the <see cref="ConcurrentExclusiveSchedulerPair.ExclusiveScheduler"/>.
+        ///
+        /// We use this as a condition of using the in-memory shared-cache sqlite DB.  This DB
+        /// doesn't busy-wait when attempts are made to lock the tables in it, which can lead to
+        /// deadlocks.  Specifically, consider two threads doing the following:
+        ///
+        /// Thread A starts a transaction that starts as a reader, and later attempts to perform a
+        /// write. Thread B is a writer (either started that way, or started as a reader and
+        /// promoted to a writer first). B holds a RESERVED lock, waiting for readers to clear so it
+        /// can start writing. A holds a SHARED lock (it's a reader) and tries to acquire RESERVED
+        /// lock (so it can start writing).  The only way to make progress in this situation is for
+        /// one of the transactions to roll back. No amount of waiting will help, so when SQLite
+        /// detects this situation, it doesn't honor the busy timeout.
+        ///
+        /// To prevent this scenario, we control our access to the db explicitly with operations that
+        /// can concurrently read, and operations that exclusively write.
+        /// 
+        /// All code that reads or writes from the db should go through this.
+        /// </summary>
+        private readonly ConcurrentExclusiveSchedulerPair _readerWriterLock = new ConcurrentExclusiveSchedulerPair();
+
+        // Inner class used so that we can generically create and pool StrongBox<TArg> instances. We
+        // use those to provide an allocation-free means of calling
+        // <c>Task.Factory.StartNew(func<object>, object, ...)</c>.
```

Comment:
@stephentoub is there any way to avoid this?  We want to call `Task.Factory.StartNew` passing a static delegate and a tuple of args for it to execute.  However, the StartNew api only allows an `object args` to be passed, which would cause allocations.  I'm working around this with pooled StrongBox instances, but i'd love if i could avoid this entirely.  Thanks! :)

Reply:
There's no StartNew overload that's generic on the argument.  Obviously if TArg is a reference type, you can just pass that through as the object.  If it's a value type, there's no StartNew API that'll let you do it without allocating: it's actually stored in the Task itself as "object".

Reply:
Ah, that's definitely a pity.  Have you guys considered a generic subclass that would allow for that extra storage without forcing it as object?

Reply:
Considered, yes, and such an approach is taken for other things in some cases.  However, it's challenging here for a few reasons, namely a) Task has constructors that take object so either the base needs to be able to store the state or it needs to be able to be stored off the Task somewhere which is likely very expensive, b) we need to ensure the state is accessible very quickly and we want to avoid things like virtuals to get at it, c) as noted we already have some task derived types and so the matrix of task hierarchies starts to become unwieldy, etc.  If you want to try your hand at what an implementation might look like, though, I'd be interested in seeing it.

That said, while this may not apply to your specific case because your API suggests you need a concrete `Task<T>`, the direction this is generally heading for folks that are trying to optimize at this level is using `ValueTask<T>` with an object that implements both `IValueTaskSource<T>` and `IThreadPoolWorkItem`.  The former allows you to provide the implementation that backs the `ValueTask<T>`, including allowing for pooled, reusable objects, and the latter in tandem with an `UnsafeQueueUserWorkItem` overload allow you to queue your own objects to the thread pool for execution.  Combined, that would mean you could avoid not only the boxing allocation for the arguments, but the allocation for the work item object itself.

Reply:
@stephentoub That makes a lot of sense.  Thanks!

I'm very curious about the valuetasksource stuff, but the complexity level seems quite high.

1. Are there good helper classes to make that stuff easier?
2. My code here ultimately goes through a ConcurrentExclusiveSchedulerPair to partition readers and writers.  I imagine i won't get a benefit from valuetask since the use of that scheduler will ultimately necessitate actual task allocations right?  or is CESP somehow amazing and can do allocation-free scheduling?

Reply:
> Are there good helper classes to make that stuff easier?

All of the core logic for implementing `IValueTaskSource<T>` is in the long-named `ManualResetValueTaskSourceCore<T>`: it provides implementations for all of the `IValueTaskSource<T>` methods that it can just delegate to.  The implementation of `IValueTaskSource<T>` then "just" needs to supply the additional logic, e.g. around how/when to reset the instance and pool it if desired.  Here's an example in the corefx tests:
https://github.com/dotnet/corefx/blob/2d69a2494b3c1602d9332bab3794e00d1b4858c7/src/System.Threading.Tasks.Extensions/tests/ManualResetValueTaskSourceTests.cs#L583-L586
Note that when Roslyn generates async enumerables, it uses MRVTSC, with the generated `IAsyncEnumerable<T>` also implementing `IValueTaskSource<T>`... that way, when `MoveNextAsync` completes asynchronously, it can do so by just returning a `new ValueTask<bool>(this, ...)`, such that no matter how many times `MoveNextAsync` is called on an enumerable, it's ammortized allocation-free.

> My code here ultimately goes through a ConcurrentExclusiveSchedulerPair to partition readers and writers. I imagine i won't get a benefit from valuetask since the use of that scheduler will ultimately necessitate actual task allocations right?

Correct.  Sorry.

> or is CESP somehow amazing and can do allocation-free scheduling?

You will still get the allocation for the Task itself, but note that CESP itself needs to schedule to its underlying scheduler, and in the case where that underlying scheduler is just the thread pool, it avoids the underlying work item allocation by itself using `IThreadPoolWorkItem`, implementing it on one of its types, which it then queues over and over any time it needs to scheduler work:
https://github.com/dotnet/coreclr/blob/98c2d5ff70059d115fe38bc6c19956a721e33558/src/System.Private.CoreLib/shared/System/Threading/Tasks/ConcurrentExclusiveSchedulerPair.cs#L339-L344